cephfs: Add CSI-Addons NetworkFence support to CephFS

riya-singhal31 opened the pull request:
-  adds client eviction to cephfs, based
on the IPs in cidr block, it evicts those IPs from
the network.

updates: https://github.com/ceph/ceph-csi/issues/3969

riya-singhal31 committed:
cephfs: adds the implementation of client eviction

this commit adds client eviction to cephfs, based
on the IPs in cidr block, it evicts those IPs from
the network.

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

riya-singhal31 committed:
cephfs: add network fence service to identity.go

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -38,6 +39,27 @@ type NetworkFence struct {
 }
 
+// ActiveClient represents the structure of an active client.
+type ActiveClient struct {
```
this struct is not used outside of this package, you do not need to export it, so call it `activeClient`.

Madhu-1 replied:
And also remove unwanted members from the struct

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	return ipCidr.Contains(ipAddress)
+}
+
+func fetchIP(clientInfo string) (string, error) {
```
make this
```go
func (ac *activeClient) fetchIP() (string, error) {
    clientInfo := ac.Inst
    ...
```

and then call
```go
clientIP, err := client.fetchIP()
```
further below.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
 
+func listActiveClients(ctx context.Context, mdsDaemonID string) ([]ActiveClient, error) {
+	cmd := []string{"tell", fmt.Sprintf("mds.%s", mdsDaemonID), "client", "ls"}
+	stdout, stdErr, err := util.ExecCommand(ctx, "ceph", cmd...)
```
make sure to leave a `FIXME:` comment to use go-ceph in the future

Madhu-1 replied:
Yes this is much required

nixpanic reviewed (commented):
An early review. Please add unit-tests for functions like `fetchIP()` so that the parsing of the client IP-address is validated.

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,86 @@
+	"context"
+	"errors"
+
+	nf "github.com/ceph/ceph-csi/internal/csi-addons/networkfence"
```
```suggestion
	nf "github.com/ceph/ceph-csi/internal/csi-addons/networkfence"
	"github.com/ceph/ceph-csi/internal/util"

	"github.com/csi-addons/spec/lib/go/fence"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
```

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,86 @@
+	"google.golang.org/grpc/status"
+)
+
+// FenceControllerServer struct of cephfs CSI driver with supported methods
```
`cephfs` to `cephFS`

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// NewFenceControllerServer creates a new FenceControllerServer which handles
+// the FenceController Service requests from the CSI-Addons specification.
```
there is a mixture of `CSI-Addons` and `CSI-addons` in multiple places, please stick to one.

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,86 @@
+	return &FenceControllerServer{}
+}
+
+func (fcs *FenceControllerServer) RegisterService(server grpc.ServiceRegistrar) {
```
missing function comment?

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	return nil
+}
+
+func isIPInCIDR(ctx context.Context, ip string, cidr string) bool {
```
we already have this for RBD right? we can just have a util function for both and use it?

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+// using client eviction.
+func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+	// Fetch active clients
+	activeClients, err := listActiveClients(ctx, "0")
```
Add a comment here what is `0` here and is it a const value which will not change ever?

riya-singhal31 replied:
Yes, we can always use mds rank 0, since all the clients have a session with rank-0
Thanks, updated!

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+				// Evict the client if it matches the CIDR block
+				err := evictCephFSClient(ctx, clientIP, "0")
+				if err != nil {
+					return fmt.Errorf("Error evicting client %s: %s\n", clientIP, err)
```
start error with small letters ad use `%w` for wrapping error

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+			// Check if the clientIP is in the CIDR block
+			if isIPInCIDR(ctx, clientIP, cidr) {
+				// Evict the client if it matches the CIDR block
+				err := evictCephFSClient(ctx, clientIP, "0")
```
what will happen for already evicted clients? are we going to see any error or it will be skipped?

riya-singhal31 replied:
Should be no-op, that is skip. I can try to test it out.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	// Parse the CIDR block
+	_, ipCidr, err := net.ParseCIDR(cidr)
+	if err != nil {
+		log.DebugLog(ctx, "Error parsing CIDR block %s: %v\n", cidr, err)
```
use error logging.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+}
+
+func evictCephFSClient(ctx context.Context, clientIP string, mdsDaemonID string) error {
+	cmd := []string{"tell", fmt.Sprintf("mds.%s", mdsDaemonID), "client", "evict", clientIP}
+	_, stdErr, err := util.ExecCommand(ctx, "ceph", cmd...)
```
use with timeout as the command might get stuck?

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +154,90 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	}
+	// Iterate through CIDR blocks and check if any active client matches
+	for _, cidr := range nf.Cidr {
+		for _, client := range activeClients {
+			clientIP, err := fetchIP(client.Inst)
+			if err != nil {
+				return fmt.Errorf("Error fetching client IP: %s\n", err)
+			}
+			// Check if the clientIP is in the CIDR block
+			if isIPInCIDR(ctx, clientIP, cidr) {
```
what if we dont find any activeClients but we still want to blocklist all the IP's in the CIDR?

riya-singhal31 replied:
Thanks @Madhu-1 , updated. 
For the IPs in CIDR which doesn't have any corresponding active clients, blocklisting like rbd does -> i.e. blocklisting the entire IP by adding to ceph blocklist.
PTAL.

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2023 The Ceph-CSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
it looks like its missing some new lines please check

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +142,134 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
 	return nil
 }
 
+func listActiveClients(ctx context.Context, mdsRank string) ([]activeClient, error) {
```
why to pass `mdsRank` as function argument its hardcoded value `0` always?

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +142,134 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+		return parts[1], nil
+	}
+
+	return "", errors.New("failed to extract IP address")
```
return default from which client we failed to fetch IP?

riya-singhal31 replied:
Just for clarity,
Did you mean to return the ID of client in place of empty string in that case?

Madhu-1 replied:
i mean to return from what we are not able to extra IP address or there should be atleast a log of clientInfo which helps of debugging

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +142,134 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+		return clientID, nil
+	}
+
+	return 0, errors.New("unable to extract client ID")
```
any more details we can return in error?

riya-singhal31 replied:
Thanks @Madhu-1, Updated.
Returning the client info too which contains the instance field of client struct, to check the missing format.

riya-singhal31 committed:
cephfs: adding unit test for fetchIP for client eviction

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

riya-singhal31 committed:
cephfs: adding unit test for fetchID

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

riya-singhal31 commented:
> An early review. Please add unit-tests for functions like `fetchIP()` so that the parsing of the client IP-address is validated.

Thanks @nixpanic, Updated.
PTAL.

riya-singhal31 committed:
cephfs: adds unit testing for fencing procedure

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,137 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+}
+
+func (ac *activeClient) fetchIP() (string, error) {
+	clientInfo := ac.Inst
```
please document (give an example) of the `.Inst` formatted string

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,137 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+}
+
+func (ac *activeClient) fetchID() (int, error) {
+	clientInfo := ac.Inst
```
Same here, how is `.Inst` formatted? A comment about the format and what part the ID is would be helpful.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,137 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+}
+
+// AddClientEviction blocks access for all the IPs in the CIDR block
+// using client eviction.
```
explain here that inactive clients will get blocklisted

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,137 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+		}
+		// if no active client found corresponding to the ip in cidr
+		// blocklisting entire ip
+		if !isFound {
```
I don't think this is complete yet.

Consider a `cidr` that contains multiple worker nodes. Maybe a single worker node has an active CephFS client, and was evicted. All other workernodes in that `cidr` need to be blocklisted.

Instead of `isFound` as bool, you could create a list of evicted clients, and skip those clients while blocklisting all IP-addresses in the CIDR.

riya-singhal31 replied:
Thanks @nixpanic for the review,
Updated, PTAL.

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -54,3 +54,43 @@ func TestGetIPRange(t *testing.T) {
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
I am not very fond of this. It would be clearer to use `tc := tt` so that `tc` is clearly an instance for the text-case. You can then use `tc` in all calls below.

nixpanic reviewed (changes requested):
Just a few minor things to improve. Generally looks pretty good to me.

riya-singhal31 committed:
cephfs: implement the logic for unfencing procedure

this commit un-blocklists the clients provided in cidr
for unfencing operation.

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

riya-singhal31 committed:
cephfs: adds unit testing for unfencing procedure

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,149 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+				}
+				log.DebugLog(ctx, "client %d has been evicted\n", clientID)
+				// add the CIDR to the list of blocklisted IPs
+				evictedIPs[cidr] = true
```
You need to track the singe client IP-address that was evicted, not the whole CIDR block. There might be other IP-addresses in the CIDR block that are not evicted, and those will need to be blocklisted.

Instead of `cidr` as key in the map, use `client.fetchIP()`.

nixpanic commented on internal/csi-addons/networkfence/fencing_test.go:
```diff
@@ -54,3 +54,43 @@ func TestGetIPRange(t *testing.T) {
+		expectedErr bool
+	}{
+		{
+			clientInfo:  "client.4305 172.21.9.34:0/422650892",
```
I guess this also needs to support IPv6 addresses? You probably should test if that works too.

riya-singhal31 replied:
Currently it works for IPv4, will add support for IPv6 in a separate new PR? can open an issue for it as a tracker.

nixpanic replied:
Ideally add support in this PR. IPv6 should not be treated differently from IPv4.
If you really want to address it in a different PR, create an issue and leave a link here.

nixpanic replied:
reported at #4192

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -287,6 +287,57 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+
+	// Parse the stdout to extract blocklist entries
+	// each entry is a separate line in the output
+	entries := strings.Split(strings.TrimSpace(stdout), "\n")
```
Give an example of the output that is expected from the ceph command. It is difficult to review this without knowing what the command returns.

Even better: move the parsing to a helper function and add a unit-test.

riya-singhal31 replied:
Thanks Niels, have added comments, explaining the workflow here and in main removeClientEviction too, PTAL.

riya-singhal31 replied:
Nothing extra is being done here for parsing, just storing them as new lines. So might not be a need for unit testing.

nixpanic replied:
Everything that can be unit-tested, should ideally be unit-tested :smile:

riya-singhal31 replied:
Yes, updated. PTAL 😄

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -287,6 +287,57 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+	// iterate through CIDR blocks and remove matching entries
+	for _, cidr := range nf.Cidr {
+		for _, entry := range blocklistEntries {
+			if strings.HasPrefix(entry, cidr) {
```
It is difficult to verify how this works. Make it easier for others (and for yourself in a few months time) to understand this by adding comments with examples of what `cidr` contains, and how `enty` is formatted.

riya-singhal31 replied:
Thanks @nixpanic , updated. PTAL

nixpanic replied:
A `cidr` in the format of `127.0.0.1:0/3710147553` is weird...

A CIDR could be `192.168.123.234/16`, a blocklisted IP in that range can be `192.168.1.2` (note that the prefix-comparing won't work here). You will need to check if the IP-address from the `entry` is included in the range of IP-addresses described by the CIDR. At the moment, the current check is not correct.

riya-singhal31 replied:
Thanks @nixpanic ,
Updated.

nixpanic reviewed (changes requested):
just needs a few more smaller improvements

riya-singhal31 committed:
cephfs: expose FenceControllerServer in driver.go to handle the requests

Signed-off-by: Riya Singhal <rsinghal@redhat.com>

Madhu-1 commented on examples/cephfs/secret.yaml:
```diff
@@ -6,12 +6,12 @@ metadata:
-  adminID: <plaintext ID>
-  adminKey: <Ceph auth key corresponding to ID above>
+  adminID: admin
+  adminKey: AQDY7RNlYlbUGhAASH0UHq8Jn4fvVmGIzsbARQ==
```
Please revert back this changes

Madhu-1 commented on internal/csi-addons/cephfs/network_fence_test.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/csi-addons/spec/lib/go/fence"
+)
+
+// TestFenceClusterNetwork is a minimal test for the FenceClusterNetwork()
+// procedure. During unit-testing, there is no Ceph cluster available, so
+// actual operations can not be performed.
+func TestFenceClusterNetwork(t *testing.T) {
+	t.Parallel()
+
+	controller := NewFenceControllerServer()
+
+	req := &fence.FenceClusterNetworkRequest{
+		Parameters: map[string]string{},
+		Secrets:    nil,
+		Cidrs:      nil,
+	}
+
+	_, err := controller.FenceClusterNetwork(context.TODO(), req)
+	assert.Error(t, err)
+}
+
+// TestUnfenceClusterNetwork is a minimal test for the UnfenceClusterNetwork()
+// procedure. During unit-testing, there is no Ceph cluster available, so actual
+// operations can not be performed.
+func TestUnfenceClusterNetwork(t *testing.T) {
+	t.Parallel()
+	controller := NewFenceControllerServer()
+
+	req := &fence.UnfenceClusterNetworkRequest{
+		Parameters: map[string]string{},
+		Secrets:    nil,
+		Cidrs:      nil,
+	}
+	_, err := controller.UnfenceClusterNetwork(context.TODO(), req)
+	assert.Error(t, err)
+}
```
what is the benefit we are getting from this unit test?

nixpanic replied:
Not much, bit it should not panic either.

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,198 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+func listActiveClients(ctx context.Context) ([]activeClient, error) {
+	// FIXME: replace the ceph command with go-ceph API in future
+	cmd := []string{"tell", fmt.Sprintf("mds.%s", mdsRank), "client", "ls"}
+	stdout, stdErr, err := util.ExecCommandWithTimeout(ctx, 5*time.Minute, "ceph", cmd...)
```
RPC timeout is 3 minutes, i would suggest keeping it for 2 minutes or even less.

riya-singhal31 replied:
Yes, will update it.
Apologies for the delay @Madhu-1 , these reviews somehow were not showing up to me before and now its showing 2 weeks ago. 
Thanks!

Madhu-1 replied:
no worries, that's my mistake, i did review but forgot to submit it 🤦🏻

riya-singhal31 commented:
Testing:                                                                                     
riyasinghal@rsinghal-mac ceph-csi % kubectl exec -c csi-addons csi-cephfsplugin-provisioner-676758bc76-7fxr4 -n rook-ceph -- csi-addons -operation NetworkFence -clusterid rook-ceph -secret default/csi-cephfs-secret -cidrs 1.1.1.1/32 -endpoint unix:///csi/csi-addons.sock 
Network fence successful%       

riyasinghal@rsinghal-mac ceph-csi % kubectl exec -c csi-addons csi-cephfsplugin-provisioner-676758bc76-7fxr4 -n rook-ceph -- csi-addons -operation NetworkUnFence -clusterid rook-ceph -secret default/csi-cephfs-secret -cidrs 1.1.1.1/32 -endpoint unix:///csi/csi-addons.sock
Network unfence successful%                  

PTAL, cc: @nixpanic , @Madhu-1 , @Rakshith-R

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -286,6 +298,113 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+	var entries []string
+	for _, line := range lines {
+		if strings.TrimSpace(line) != "" && !strings.HasPrefix(line, "listed") {
+			entries = append(entries, line)
```
Are complete entry lines used anywhere? If not, you could call `fetchIPfromEntry()` here and add the IP-address to the `[]string` that is returned.

If that is an option, you could rename `getBlocklistEntries()` to `getBlocklistIPs()` and make the code a little simpler.

riya-singhal31 replied:
Thanks, @nixpanic 
Updated, PTAL.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -286,6 +298,113 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+			// fetch IP from entry
+			entryIP, err := fetchIPfromEntry(entry)
+			if err != nil {
+				return fmt.Errorf("error fetching client IP: %w", err)
```
Consider including the `entry` contents in the error message (with `%q`). It'll make troubleshooting and debugging much easier.

nixpanic reviewed (dismissed):
Looks complete to me. Left a few minor comments that could improve the code a little more.

riya-singhal31 commented:
> Looks complete to me. Left a few minor comments that could improve the code a little more.

Thanks @nixpanic , Updated! 
Requesting for a re-look.

Madhu-1 commented on internal/csi-addons/cephfs/network_fence_test.go:
```diff
@@ -0,0 +1,60 @@
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
```
no new line is required

Madhu-1 commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -132,6 +145,268 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+			if evictedIPs[host] {
+				continue
+			}
+			err = nf.addCephBlocklist(ctx, host, false)
+			if err != nil {
+				return err
+			}
```
i think this can lead to a problem for below case

* Create RBD NetworkFence CR for one IP 10.10.10.10
* Created CephFS NetworkFence CR for IP range but above IP comes in the Range
* Delete the CephFS Network Fence CR to unblocklist the IP

Now the expectation is the IP that is used in RBD NetworkFence is also blocklisted but that's not the case anyone isn't it?

nixpanic replied:
Yes, that looks possible.

I do not know if it is possible to see if a blocklisted client is CephFS or RBD. If that is returned (or resolvable otherwise) from listing the blocklisted client output, a check for the type can be added. Otherwise, we'll need to com up with some other way of tracking/identifying.

riya-singhal31 replied:
The blocklisted output doesn't have anything to distinguish between two
ex- $ ceph osd blocklist ls
listed 1 entries
127.0.0.1:0/3710147553 2018-03-19 11:32:24.716146

So, one way could be to add a field in NetworkFence struct to track the type of network fence (RBD or CephFS). When creating a NetworkFence object, we can set this field based on the network fence type, and then while evicting we can check if this has only cephfs and not rbd then evict. 
WDYT @Madhu-1 , @nixpanic

Madhu-1 replied:
i dont think we can make any specific change related to cephcsi in the Spec/API. we can decide what to support and what not to support from the cephcsi point. if you are talking about faster recovery of cephfs volume and if that is the whole responsibility of this API you can just document here, use RBD network fence to fence the whole cluster IP and use cephfs network fence to fence any active client which helps to evict only active client, if there are no active client, just return the success or if there any active client in the range just blocklist them and return success. @nixpanic thoughts?

nixpanic replied:
I guess the behavior can be documented as some kind of corner-case. Most workflows will NetweorkFence a node (or whole cluster) for both RBD and CephFS. If there is an issue with the node, neither RBD and CephFS is expected to function. Unfencing one of the protocols suggests the node is expected to be recovered, so both CephFS and RBD are expected to work again too.

We can improve this in a future version of the CSI-Addons Spec, I guess. A NetworkFence operation could pass, and update a NetworkFence.Context, similar to what a PersistentVolume.Context has. In a Kubernetes environment, a NetworkFenceClass might even be appropriate, it could link to secrets and contain driver specific options.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -286,6 +298,115 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+	return blocklistIPs, nil
+}
+
+func (nf *NetworkFence) removeBlocklistEntry(ctx context.Context, entry string) error {
```
hmm, is the function really different from `removeCephBlocklist()` ?

https://github.com/ceph/ceph-csi/blob/devel/internal/csi-addons/networkfence/fencing.go#L178

Maybe the same function can be used?

riya-singhal31 replied:
Right Niels, thanks!
It's the same computation, for the whole un-blocklisting the same logic can be used.

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -286,6 +298,115 @@ func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+
+// RemoveClientEviction unblocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func (nf *NetworkFence) RemoveClientEviction(ctx context.Context) error {
```
This might be similar a `RemoveNetworkFence()` from https://github.com/ceph/ceph-csi/blob/devel/internal/csi-addons/networkfence/fencing.go#L202

riya-singhal31 commented:
Updated the PR with all the reviews,
Please take a look! 
@nixpanic , @Madhu-1 

Thanks!

nixpanic reviewed (dismissed):
Looks good to me, thanks!

Madhu-1 commented on internal/csi-addons/cephfs/network_fence.go:
```diff
@@ -0,0 +1,120 @@
+
+	cr, err := util.NewUserCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
```
Internal error or invalid? in UnfenceClusterNetwork its Invalid and here its Internal.

riya-singhal31 replied:
Thanks Madhu, should be InvalidArgument, updated. PTAL.

Madhu-1 reviewed (commented):
LGTM, one small change is required.

Madhu-1 commented:
@riya-singhal31 instead of a new commit, please amend this change to the commit in which this was implemented

riya-singhal31 commented:
> @riya-singhal31 instead of a new commit, please amend this change to the commit in which this was implemented

Done!

nixpanic commented:
@mergifyio rebase
